src/kernel/kernel.h

```diff
@@ -89,6 +89,7 @@ static inline __attribute__((always_inline)) void cpu_setgs(int cpu)
     u64 addr = u64_from_pointer(cpuinfo_from_id(cpu));
     write_msr(KERNEL_GS_MSR, 0); /* clear user GS */
     write_msr(GS_MSR, addr);
+    write_msr(TSC_AUX_MSR, cpu);    /* used by vdso_getcpu() */
```

Comment:
I'm not sure if a write to a non-present MSR would cause a fault, but it might be safer to first check for presence of rdtscp / IA32_TSC_AUX.

From intel SDM, vol 4, 2-46 on IA32_TSC_AUX:

"If CPUID.80000001H:EDX[27] = 1 or CPUID.(EAX=7,ECX=0):ECX [bit 22] = 1"

and in Vol 2A, 3-214, for cpuid EAX=80000001H:

"[EDX] Bit 27: RDTSCP and IA32_TSC_AUX are available if 1."

Reply:
OK, I made the MSR write conditional to __vdso_dat->platform_has_rdtscp.
The cpu_setgs() function is being replaced by a more generic cpu_init() function (moved from kernel.h to arch-specific code, which also solves issues due to double #includes of vdso.h), where TSC_AUX is initialized (if RDTSCP is supported).
Since the platform_has_rdtscp flag is initialized in init_clock(), I moved the init_clock() call to before the call to init_kernel_contexts() (which indirectly calls cpu_init() for CPU 0).